edx/edx-enterprise
# ENT-220/ENT-223 Collect learner completion data for SAP SuccessFactors

**Description:** Adds a management command and updates supporting classes to collect learner completion data records, and prepare them for submission to the SAP SuccessFactors integration channel.

**JIRA:** [ENT-220](https://openedx.atlassian.net/browse/ENT-220), [ENT-223](https://openedx.atlassian.net/browse/ENT-223)

**Dependencies:** 

* https://github.com/edx/edx-platform/pull/14563
* https://github.com/edx/edx-enterprise/pull/61 (merged here)

**Sandbox:**

* LMS: https://ent-pr59.sandbox.opencraft.hosting/
* Studio: https://studio-ent-pr59.sandbox.opencraft.hosting/

Also created an edX Sandbox: https://ent-220-ent-223-pomegranited.sandbox.edx.org
I've added github keys for @haikuginger , @brittneyexline , @mattdrayer , so you can shell in as `pomegranited` and test the management command.  

**Merge deadline:** 13 March 2017

**Installation instructions:** 

1. Ensure edx-platform is running with changes from https://github.com/edx/edx-platform/pull/14563.
1. Install this branch's latest commit on your edxapp devstack:

    ```
    pip uninstall -y edx-enterprise && pip install "git+https://github.com/open-craft/edx-enterprise.git@jill/sap_learner_completion#egg=edx-enterprise"
    ```
1. You may have to run data migrations:

   ```
    ./manage.py lms migrate --settings=devstack
   ```
1. Restart the LMS without clobbering the installed `edx-enterprise`:

    ```
    NO_PREREQ_INSTALL=1 paver run_all_servers --fast
    ```
1. Set up a [TestShib SAML integration](http://testshib.org) with [these instructions](http://edx.readthedocs.io/projects/edx-installing-configuring-and-running/en/latest/configuration/tpa/tpa_SAML_SP.html).

**Testing instructions:**

1. In Studio, create a new course.
1. From the Admin, create a course mode for the course under `Course_Modes › Course modes`.  Choose a mode other than Audit.
1. In Studio,  navigate to the Settings > Certificates page, and create a certificate.  Don't worry about adding signatories.  Activate the certificate.
1. Run Testing instruction steps 1-4 from https://github.com/edx/edx-platform/pull/14400.  Use a new learner, so that it can be registered via TestShib.  Enrol the new learner in your course.
1. Link the `EnterpriseCustomer` to the TestShib SAML integration as the EC's `Identity Provider`.
1. From the Admin, add an [Enterprise SAP SuccessFactors Integration › Sap success factors enterprise customer configurations](http://localhost:8000/admin/sap_success_factors/sapsuccessfactorsenterprisecustomerconfiguration/) for the `EnterpriseCustomer`.  The only crucial field is `enabled=True`.  The other fields do't matter yet, because no actually communication is being done.
1. Run the new management command.  No output expected, unless any of your enrolled learners have granted consent.

    ```
    # For all enterprise customers
    ./manage.py lms transmit_learner_data  --settings=devstack

    # Or just one.
    ./manage.py lms transmit_learner_data --enterprise_customer=<enterprise_customer.uuid>  --settings=devstack
    ```
1. Register your enrolled learner via TestShib, and grant data sharing consent.
1. Re-run the `transmit_learner_data` management command to see the learner data record logged. Note that it reports `courseCompleted: "false"`.
1. Generate a certificate for your course.  

    ```
    ./manage.py lms ungenerated_certs --course="<course_id>" --settings=devstack
    ```
    *Note*: I got a lot of `VersionConflict` errors when I ran this, which didn't seem to matter.  And the log line that indicates the certificate has been generated may say something like "Student 5 has certificate status 'unavailable'".  But that seems to be ok too?
1. Re-run the `transmit_learner_data` management command, and check that the output now says `courseCompleted: "true"`, and the `grade` field has a value.

**Author notes and concerns:**

* `grade` field value may change, depending on discussion.  Currently using the `certificate.grade`, but may provide `Pass/Fail` instead.
* What to do about audit modes, in self-paced courses with no certificates to indicate completion and/or final grade?  Currently, we can only send learner completion data when a certificate has been issued.

**Reviewers:**

- [x] @brittneyexline 

**Merge checklist:**

  **DO NOT** just add dependencies to `requirements/*.txt` files.
- [x] Version bumped - to `0.27.0`
- [ ] Translations are updated - [see comment](https://github.com/edx/edx-enterprise/pull/59#issuecomment-285582381)

**Post merge:**
